src/day16.scala

```diff
@@ -0,0 +1,201 @@
+  def version: Int
+  def typeId: Int
+  def value: Long
+
+sealed trait OperatorPacket extends Packet:
+  def version: Int
+  def typeId: Int
+  def value: Long
+  def exprs: List[Packet]
+
+sealed trait BinaryOperatorPacket extends OperatorPacket:
+  def version: Int
+  def typeId: Int
+  def value: Long
+  def rhs: Packet
+  def lhs: Packet
+  def exprs = List(lhs, rhs)
+
+case class SumPacket(version: Int, exprs: List[Packet]) extends OperatorPacket:
+  val typeId = 0
+  def value = exprs.map(_.value).sum
+
+case class ProductPacket(version: Int, exprs: List[Packet])
+    extends OperatorPacket:
+  val typeId = 1
+  def value = exprs.map(_.value).reduce(_ * _)
+
+case class MinimumPacket(version: Int, exprs: List[Packet])
+    extends OperatorPacket:
+  val typeId = 2
+  def value = exprs.map(_.value).min
+
+case class MaximumPacket(version: Int, exprs: List[Packet])
+    extends OperatorPacket:
+  val typeId = 3
+  def value = exprs.map(_.value).max
+
+case class LiteralPacket(version: Int, value: Long) extends Packet:
+  val typeId = 4
+
+case class GreaterThanPacket(version: Int, lhs: Packet, rhs: Packet)
+    extends BinaryOperatorPacket:
+  val typeId = 5
+  def value = if lhs.value > rhs.value then 1 else 0
+
+case class LesserThanPacket(version: Int, lhs: Packet, rhs: Packet)
+    extends BinaryOperatorPacket:
+  val typeId = 6
+  def value = if lhs.value < rhs.value then 1 else 0
+
+case class EqualsPacket(version: Int, lhs: Packet, rhs: Packet)
+    extends BinaryOperatorPacket:
+  val typeId = 7
+  def value = if lhs.value == rhs.value then 1 else 0
+
+
+/*
+ * Parsing of packets
+ */
+
+type BinaryData = List[Char] 
+
+inline def toInt(chars: BinaryData): Int =
+  Integer.parseInt(chars.mkString, 2)
+
+inline def toLong(chars: BinaryData): Long =
+  java.lang.Long.parseLong(chars.mkString, 2)
+
+@tailrec
+def readLiteralBody(tail: BinaryData, numAcc: BinaryData): (Long, BinaryData) =
+  val (num, rest) = tail.splitAt(5)
+  if num(0) == '1' then readLiteralBody(rest, numAcc.appendedAll(num.drop(1)))
+  else
+    val bits = numAcc.appendedAll(num.drop(1))
+    (toLong(bits), rest)
+end readLiteralBody
+
+def readOperatorBody(current: BinaryData): (List[Packet], BinaryData) =
+  val (lenId, rest) = current.splitAt(1)
+
+  @tailrec
+  def readMaxBits(
+      current: BinaryData,
+      remaining: Int,
+      acc: List[Packet]
+  ): (List[Packet], BinaryData) =
+    if remaining == 0 then (acc, current)
+    else
+      val (newExpr, rest) = decodePacket(current)
+      readMaxBits(rest, remaining - (current.size - rest.size), acc :+ newExpr)
+
+  @tailrec
+  def readMaxPackages(
+      current: BinaryData,
+      remaining: Int,
+      acc: List[Packet]
+  ): (List[Packet], BinaryData) =
+    if remaining == 0 then (acc, current)
+    else
+      val (newExpr, rest) = decodePacket(current)
+      readMaxPackages(rest, remaining - 1, acc :+ newExpr)
+
+  lenId match
+    // read based on length
+    case List('0') =>
```

Comment:
Have you considered
```scala
if lenId(0) == '0' then
```
like you did in `readLiteralBody`?

Reply:
Changed, good idea. Wasn't too happy with that code.